### DISCUSSION TITLE
How does the step ratio of text encoder training choose which epochs to train text encoder?

### DISCUSSION THREAD

**Discussion by minienglish1** [d8ahazard/sd_dreambooth_extension#1048] [Q&A]
I looked through the code, and it looks like:
1) Multiply text ratio by total epochs = text encoder epochs
2) Train text encoder for set number of epochs
3) When finished training text encoder, it continues training the unet until total epochs is reached

Is this correct?

If so, how does this function with resuming training?

If not using a checkpoint, does it just reset and train the text encoder again via the above formula?
If using a checkpoint, does it account for text encoder epochs already trained?

If this is correct, then my prior testing has some issues.  I just assumed it would randomly choose which epochs to train the text encoder and I wasn't using checkpoints.

In addition, how does freeze CLIP normalization affect all of this?  The only code I code find about it was:

train_tenc = epoch < text_encoder_epochs
            if stop_text_percentage == 0:
                train_tenc = False
            if not args.freeze_clip_normalization:
                text_encoder.train(train_tenc)
            else:
                text_encoder.eval()

I'm not sure what the text_ecoder.eval() does.

**Comment by Arilziem** [comment]
I did not write any of the code but this is how I read it:

1. Text encoder ratio multiplied with total epochs (and rounded) = text_encoder_epochs (same as you)
2. Every epoch, text encoder is trained after the unet is trained (for one epoch), until `text_encoder_epochs` is reached
3. After `text_encoder_epochs` is reached, only unet is trained

Resuming works the same as with unet. Prior number of epochs is ignored, you always train a session according to your settings. So every time you start a training, it will again train unet, then text encoder, then unet, then ... until text encoder ratio has been reached.

I'm not a computer scientist so I don't understand the effect of this, but eval vs train seems to change what parts of the CLIP model are trained. What confuses me though is that [per documentation](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.eval), `.eval()` is  equivalent to `.train(false)`~, so it seems like using `freeze_clip_normalization` is doing the same as setting the ratio to 0?~ No sorry, I ignored other code parts where stuff is being loaded and so on, depending on whether the text encoder ratio is 0 or not. So yeah, I can't answer that part about the consequences.

**Reply by minienglish1** [reply]
Thanks for the help.

I just tried loading from a checkpoint: it does not ignore prior epochs:
1) Setup for 100 epochs
2) Trained for 75 epochs, then closed the cmd window.
3) Then resumed training from the 75 epoch checkpoint.
If I setup to train for 75 or less epochs, it won't train.  Setup for 100 epochs, and it trained only the remaining 25 epochs, then finished.  So, loading from a checkpoint does remember the prior training.  Not using a checkpoint does not.  Good to know.

Regarding the freeze_clip_normalization, that's what was confusing me too.  I also thought ".eval() is equivalent to .train(false)."  

Thanks for the info.  I went back to my references and reviewed normalization in machine learning.  I'll go back through the code and re-read, and slowly figure out what's going on.  Then do some tests to see if it helps my model or not.

**Reply by Arilziem** [reply]
Good to know (if a bit confusing) that checkpoints behave differently from directly training on the main model, thanks. It should also correctly resume the learning rate for stuff like cosine_annealing then. Will have to try that out next time.

**Reply by minienglish1** [reply]
Let me know if it works for resuming learning rates.  That would be very useful.

**Reply by minienglish1** [reply]
Sorry, used the wrong word.  "Snapshots", not "checkpoints", save prior training data.

**Reply by Arilziem** [reply]
Yes, snapshots :3 Sadly can't test the LR thing right now as for some reason I always OOM at the moment (regardless of git revision) - unless I use LORA, which does not do snapshots. Will do once I figured out what is going on.

**Reply by DuroCuri** [reply]
The default parameter of text-encoder training is 0. As I watched some youtuber said, the text-encoder is the result will more like the tags that is already in this ckpt. The rest epochs after text-encoder is training U-Net, with will make the result more like these pictures in dataset. text-encoder will make result more overfitting(过拟合 in chinese), arise U-Net will fix it.
